`Plagiarism checks`: Add plagiarism cases for exams

akesfeden opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] (Indirectly included at the loaded module) I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
This PR extends the plagiarism features present for course exercises to support exam exercises as well. This allows the instructors to conduct plagiarism checks and to set verdict for exams and allows students to see the plagiarism cases related to their exam submissions.

### Description
<!-- Describe your changes in detail -->
New endpoint to list plagiarism cases for an exam for instructors are added. The existing components at client are extended toı support exam exercises.

Additionally, this PR fixes an issue common in both course and exam plagiarism cases where plagiarism case data is not updated when notification for a different plagiarism case is clicked when on the plagiarism case student detail view. Example: When a student clicks the notification for `/courses/1/plagiarism-cases/2` when they are at `/courses/1/plagiarism-cases/1` the page is not updated. 

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->
Prerequisites:
- 1 Instructor
- 2 Students
- 1 Exam
  - 1 Text Exercise with plagiarism case
  - 1 Modeling Exercise with plagiarism case
  - 1 Programming Exercise with plagiarism case

1. Log in to Artemis
2. Navigate to Course Administration
3. Make two similar submissions with both students.
4. As the instructor, set the end of working time of the exam to now and run plagiarism check from Exam Detail page > Exercise Groups > Detail page of one of the text/modeling/programming exercises > Check Plagiarism page.
5. Notify a student about the plagiarism.
6. With the notified student, ensure they **cannot access any sensitive data** for the related exam or the plagiarism case.
7. With the notified student, check if the exam result page shows the red `Plagiarism Case` button near the affected exercise.
8. As the instructor, set a verdict.
9. With the notified student, check if the plagiarism page shows the verdict (you can navigate there from the received notification or the `Plagiarism Case`/`Resolved Plagiarism Case` button near the affected exercise.
10. Repeat steps above for one of each text/modeling/programming exercise types in an exam.
11. Ensure that exam conduction is not affected and still working as before.
12. Access to repos belonging to programming exercises in exams working as before except when the student is notified for a relevant plagiarism case, in this case the student should be able to access all repos involved in that plagiarism case to compare files.

Navigation from notification bugfix:
1. With a student, navigate to any plagiarism detail that they have access to (or type `/courses/:courseId/plagiarism-cases/:plagiarismCaseId` with some valid id values)
2. At the notification panel, click on a plagiarism case notification for a different plagiarism case.
3. Ensure that the page is updated correctly.

### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code as well as the functionality (= manual test) needs to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->

#### Code Review
- [x] Review 1
- [ ] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
<!-- Please add the test coverages for all changed files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan. -->
<!-- Lines are the main reference but a significantly lower branch percentage can indicate missing edge cases in the tests. -->
| Class/File | Branch | Line |
|------------|-------:|-----:|
| Post.java | 65% | 86% |
| PlagiarismCaseResource.java | 75% | 95% |
| plagiarism-cases-instructor-view.component.ts | 76.08% | 100% |
| plagiarism-cases.service.ts | 71.42% | 100% |
| plagiarism-case-instructor-detail-view.component.ts | 89% | 98.79% |
| plagiarism-case-student-detail-view.component.ts | 75% | 95.83% |
| exam-detail.component.ts | 75% | 96.22% |
| exam-participation-summary.component.ts | 79.77% | 93.06% |

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

![image](https://user-images.githubusercontent.com/18427209/185949979-4bb8060b-b79d-4105-b269-2a44efde1e46.png)

![image](https://user-images.githubusercontent.com/18427209/186154361-801cbff0-5459-4907-912c-c050fcd58dbd.png)

![image](https://user-images.githubusercontent.com/18427209/185950121-96658fd5-320f-458a-9414-60206ea00e9f.png)

![image](https://user-images.githubusercontent.com/18427209/185950574-730ee52f-6e21-4af3-9958-fc7c356b5d18.png)

![image](https://user-images.githubusercontent.com/18427209/185950981-d480a7eb-ebf9-4229-a0c2-b7e96d5d9f9e.png)

b-fein commented on src/test/java/de/tum/in/www1/artemis/PlagiarismCaseIntegrationTest.java:
```diff
@@ -171,6 +180,51 @@ void testGetPlagiarismCaseForInstructor() throws Exception {
+                PlagiarismCase.class);
+        assertThat(plagiarismCasesResponse).as("should get exam plagiarism cases for instructor").isEqualTo(examPlagiarismCases);
+        for (var submission : plagiarismCasesResponse.get(0).getPlagiarismSubmissions()) {
+            assertThat(submission.getPlagiarismComparison().getPlagiarismResult().getExercise()).as("should prepare plagiarism case response entity").isEqualTo(null);
```
```suggestion
            assertThat(submission.getPlagiarismComparison().getPlagiarismResult().getExercise()).as("should remove unneeded elements from the response").isNull();
```
I think the `as()` description could be a bit more clear. Additionally, the more specific assertion could be used.

The same applies to the the other two lines below as well.

b-fein commented on src/main/java/de/tum/in/www1/artemis/web/rest/plagiarism/PlagiarismCaseResource.java:
```diff
@@ -66,6 +69,31 @@ public ResponseEntity<List<PlagiarismCase>> getPlagiarismCasesForCourseForInstru
+    public ResponseEntity<List<PlagiarismCase>> getPlagiarismCasesForExamForInstructor(@PathVariable long courseId, @PathVariable long examId) {
+        log.debug("REST request to get all plagiarism cases for instructor in exam with id: {}", examId);
+        Course course = courseRepository.findByIdElseThrow(courseId);
+        if (!authenticationCheckService.isAtLeastInstructorInCourse(course, userRepository.getUserWithGroupsAndAuthorities())) {
```
```suggestion
        if (!authenticationCheckService.isAtLeastInstructorInCourse(course, null)) {
```
There is no benefit of fetching the user here if it is not used anywhere else afterwards.

b-fein commented on src/main/java/de/tum/in/www1/artemis/web/rest/plagiarism/PlagiarismCaseResource.java:
```diff
@@ -66,6 +69,31 @@ public ResponseEntity<List<PlagiarismCase>> getPlagiarismCasesForCourseForInstru
+        }
+
+        var plagiarismCases = plagiarismCaseRepository.findByExamIdWithPlagiarismSubmissionsAndComparison(examId);
+        plagiarismCases.forEach(plagiarismCase -> {
+            if (plagiarismCase.getExercise().getExerciseGroup().getExam().getCourse().getId() != courseId) {
+                throw new AccessForbiddenException("Exercise with id " + plagiarismCase.getExercise().getId() + " is not related to the given course id " + courseId);
+            }
+        });
```
Would it be enough here to check that the exam with the given `examId` belongs to the course? If the exam belongs to a course, then all of its exercises should as well, I assume.

akesfeden replied:
I modified it to only check the exam of the first exercise since others should be the same. How does it look like now?

b-fein commented on src/main/java/de/tum/in/www1/artemis/web/rest/plagiarism/PlagiarismCaseResource.java:
```diff
@@ -147,6 +175,31 @@ public ResponseEntity<PlagiarismCaseInfoDTO> getPlagiarismCaseForExerciseForStud
+        var plagiarismCases = plagiarismCaseRepository.findByStudentIdAndExerciseIdsWithPost(user.getId(), exerciseIds);
+        var plagiarismCaseInfoDTOs = plagiarismCases.stream()
+                // the following line is already checked in the SQL statement, but we want to ensure it 100%
+                .filter(plagiarismCase -> plagiarismCase.getPost() != null).collect(Collectors.toMap(plagiarismCase -> plagiarismCase.getExercise().getId(),
```
Can there be multiple cases per exercise here? In that case in the mapping of `exerciseId -> case` would contain only the one whichever came last in the query response.

akesfeden replied:
No multiple cases per exercise with the same student id is not expected here. We can confirm that by looking at the corresponding singular method [findByStudentIdAndExerciseIdWithPost](https://github.com/ls1intum/Artemis/blob/feature/plagiarism-check-for-exams/src/main/java/de/tum/in/www1/artemis/repository/plagiarism/PlagiarismCaseRepository.java#L48-L55) returns an `Optional<PlagiarismCase>`.

b-fein replied:
Okay, in that case the name `plagiarismCases` is a bit misleading. It may also help to explicitly specify the type here instead of `var`.

akesfeden replied:
I renamed the variable and used explicit types there.

In the case you described, there is still 1 `(exerciseId, studentId)` pair created but 2 `PlagiarismSubmission` instances are linked to that. It looks like this in the instructor detail view (student view is also similar).

![image](https://user-images.githubusercontent.com/18427209/186403352-d9a320bf-f444-4b91-b759-51ac49219629.png)

![image](https://user-images.githubusercontent.com/18427209/186403361-a993510f-7010-4300-8bf4-451f0e21a776.png)

b-fein reviewed (commented):
Code review of server code. I’ve got a few questions annotated inline.

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/web/rest/plagiarism/PlagiarismCaseResource.java:
```diff
@@ -66,6 +69,33 @@ public ResponseEntity<List<PlagiarismCase>> getPlagiarismCasesForCourseForInstru
+    public ResponseEntity<List<PlagiarismCase>> getPlagiarismCasesForExamForInstructor(@PathVariable long courseId, @PathVariable long examId) {
+        log.debug("REST request to get all plagiarism cases for instructor in exam with id: {}", examId);
+        Course course = courseRepository.findByIdElseThrow(courseId);
+        if (!authenticationCheckService.isAtLeastInstructorInCourse(course, null)) {
+            throw new AccessForbiddenException("Only instructors of this course have access to its plagiarism cases.");
+        }
```
```suggestion
        authenticationCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.INSTRUCTOR, course, null);
```

akesfeden commented on src/main/java/de/tum/in/www1/artemis/web/rest/plagiarism/PlagiarismCaseResource.java:
```diff
@@ -66,6 +69,33 @@ public ResponseEntity<List<PlagiarismCase>> getPlagiarismCasesForCourseForInstru
+            var plagiarismCase = plagiarismCases.get(0);
+            var exam = plagiarismCase.getExercise().getExerciseGroup().getExam();
+            if (exam.getCourse().getId() != courseId) {
+                throw new AccessForbiddenException("Exam with id " + exam.getId() + " is not related to the given course id " + courseId);
```
You are right, I changed it accordingly.

ge65cer reviewed (changes requested):
Tested during testing session on TS5. I have the following comments:
- Problems with old plagiarism cases when the exam was reset that should be investigated
- Do all the verdicts make sense for an exam, or should we maybe limit them to only plagiarism/non-plagiarism?
- The plagiarism case page for programming exercises can not load the files from the repositories (probably because the repository gets locked in exams?)
- Do you plan to introduce a page listing all plagiarism cases for a student? (Idea of @b-fein)
- Button placement in the exam submission page should be optimized

<img height="100" alt="Screenshot 2022-08-24 at 15 35 14" src="https://user-images.githubusercontent.com/102237363/186432672-f48d0b58-bb0c-4f23-a9f6-fa4765ce1bc7.png"><img height="100" alt="Screenshot 2022-08-24 at 15 35 23" src="https://user-images.githubusercontent.com/102237363/186432691-83ddf94b-d8e5-407f-a0dd-10e640cf3c85.png">

b-fein reviewed (changes requested):
Tested in the testing session. Works well apart from three issues
- resetting an exam doesn’t clear existing cases
- for programming exercises I can’t see the files of my or the other submissions in the case overview as I’m not allowed to access that endpoint
- ~~the calculations for the final points when using the grade bonus type were wrong (e.g., 3 +/- 0.9 = 5).~~ wrong PR, comment applies to #5222

akesfeden commented:
@ge65cer @b-fein About the exam reset issue: I reproduced it locally and it turns out this was because of a foreign key in `ModelElement` referencing a `Submission` instance that should be deleted with the reset operation. I checked that `ModelElement`s are not used in plagiarism checks, they are used by model training with Compass.

Since the cause of the issue is not related to plagiarism checks and since properly testing changes to the exam operations are extra important and I am not very familiar with Compass, I think it is better to address this problem in its own PR. I will create a new issue with more details and relevant stack trace.

akesfeden commented:
@ge65cer @b-fein I fixed the issue with programming exercise files (access forbidden) and added some server tests for it, you can confirm it by navigating to the plagiarism case view with the same students you tested on TS5.

akesfeden commented:
@ge65cer @b-fein I fixed the issue with programming exercise files (access forbidden) and added some server tests for it, you can confirm it by navigating to the plagiarism case view with the same students you tested on TS5.

akesfeden commented:
@ge65cer 
> Problems with old plagiarism cases when the exam was reset that should be investigated

Explained in https://github.com/ls1intum/Artemis/pull/5620#issuecomment-1229520204

> Do all the verdicts make sense for an exam, or should we maybe limit them to only plagiarism/non-plagiarism?

As far as I know, there are no plans to disable some verdicts for exams but we can discuss it with @krusche if you suggest warning and point deduction should be disabled.
 
> The plagiarism case page for programming exercises can not load the files from the repositories (probably because the repository gets locked in exams?)

Explained in https://github.com/ls1intum/Artemis/pull/5620#issuecomment-1229521058

> Do you plan to introduce a page listing all plagiarism cases for a student? (Idea of @b-fein)

Sounds useful, however this is also not present for course exercise plagiarism cases as well so they may be both added together in a future PR.

> Button placement in the exam submission page should be optimized

Addressed in https://github.com/ls1intum/Artemis/pull/5620/commits/f4f6d87a16eb943d26261df6ed13e69c5017ec71

akesfeden commented:
@b-fein 

> resetting an exam doesn’t clear existing cases

Explained in https://github.com/ls1intum/Artemis/pull/5620#issuecomment-1229520204

> for programming exercises I can’t see the files of my or the other submissions in the case overview as I’m not allowed to access that endpoint

Explained in https://github.com/ls1intum/Artemis/pull/5620#issuecomment-1229521058

> the calculations for the final points when using the grade bonus type were wrong (e.g., 3 +/- 0.9 = 5).

I think this comment belongs to https://github.com/ls1intum/Artemis/pull/5222, right?

ge65cer reviewed (approved):
All change requests were addressed, I re-tested and found no issues. The data transferred does not seem to include any sensitive information about the exam. Thereforre, I approve.

<img width="200" alt="Screenshot 2022-08-30 at 20 24 01" src="https://user-images.githubusercontent.com/102237363/187514308-14395a68-23f8-4cba-9a9e-55d11fccd6a6.png">

ge65cer commented:
Found a small responsive design issue, the button placement could be improved for smaller screens. You may do that here or in a follow-up.

<img width="300" alt="Screenshot 2022-08-30 at 20 27 13" src="https://user-images.githubusercontent.com/102237363/187515217-1965318a-3464-41b9-b5ab-454bcf9adbeb.png"> <img width="300" alt="Screenshot 2022-08-30 at 20 29 15" src="https://user-images.githubusercontent.com/102237363/187515214-644c3353-dfc2-494b-9aeb-0f0131760a1e.png">

akesfeden commented:
@ge65cer Thanks for the review! I checked the responsiveness issue with the button, it seems the overlapping behavior with text exercises happens when the problem statement is empty. When I added it is displayed better and likewise when I removed the problem statement in `develop` branch, "View Results" button behaved similarly as well.
![image](https://user-images.githubusercontent.com/18427209/187521159-e94735b9-d0db-4c59-8937-ca83bd01b47d.png)
![image](https://user-images.githubusercontent.com/18427209/187521619-50b66114-b307-430d-b5e3-f1007cd0231a.png)

For the programming exercises, a horizontal scrollbar appears when it is shifted too much so the user is able to see the contents.

According to the findings above, it seemed that this responsiveness issue does not cause a critical problem so it makes sense to address these in a future PR as you already mentioned.

ge65cer commented on src/main/webapp/app/course/plagiarism-cases/instructor-view/detail-view/plagiarism-case-instructor-detail-view.component.ts:
```diff
@@ -193,7 +197,8 @@ export class PlagiarismCaseInstructorDetailViewComponent implements OnInit, OnDe
             student: studentName,
             exercise: exerciseTitle,
-            course: courseTitle,
+            inCourseOrExam: 'artemisApp.plagiarism.plagiarismCases.notification.' + (belongsToExam ? 'inExam' : 'inCourse'),
```
I think you're missing a call to the translation service here.

akesfeden replied:
Thanks for noticing! I fixed it at https://github.com/ls1intum/Artemis/pull/5620/commits/3671d7574d8d9c3a514af403d21939aaea4a6712

stefanwaldhauser commented on src/main/webapp/app/course/plagiarism-cases/student-view/detail-view/plagiarism-case-student-detail-view.component.ts:
```diff
@@ -32,18 +32,34 @@ export class PlagiarismCaseStudentDetailViewComponent implements OnInit, OnDestr
+        }).subscribe(({ ancestorParams, params }: { ancestorParams: Params; params: Params }) => {
+            this.courseId = ancestorParams.courseId;
+            this.plagiarismCaseId = Number(params.plagiarismCaseId);
+            if (this.plagiarismCase?.id === this.plagiarismCaseId) {
```
❓Question: Just to confirm, this is if the plagiarismCase is already loaded right? How could this happen that it is loaded before the ngOnInit?

akesfeden replied:
Yes, you are right. The aim is fixing the issue where the plagiarism case is not updated when a notification was clicked (or when the route changes without a page load for any other potential reason). It also works as expected for the initial load because before any `plagiarismCase` is loaded, the `if` condition becomes `undefined === (someNumber)` so it still sends the request. The client test added in https://github.com/ls1intum/Artemis/pull/5620/commits/a30a65d80b7d26a42a536a4df42a59310f53463d and the test just above the new one that is partially modified may explain the purpose better along with the commit message.

stefanwaldhauser commented on src/main/webapp/app/exam/participate/summary/exam-participation-summary.component.ts:
```diff
@@ -28,11 +31,13 @@ export class ExamParticipationSummaryComponent implements OnInit {
      * Current student's exam.
      */
     private _studentExam: StudentExam;
+    plagiarismCaseInfos: { [exerciseId: number]: PlagiarismCaseInfo } = {};
```
😍Cool: I love the use of indexable type here. Will have to remember that! Gj

akesfeden replied:
Thank you very much! 😍

stefanwaldhauser reviewed (approved):
Good Job! Some requested changes ❌, ideas 💡 and questions ❓. Only changes ❌ need to be addressed for approval.

b-fein reviewed (dismissed):
Code re-review: looks good. :+1:

ge65cer reviewed (approved):
Re-tested, found no more issues. Approved!

minrows reviewed (approved):
Tested on ts1 with the author and everything works as described. LGTM 👍

jpbernius commented on src/main/java/de/tum/in/www1/artemis/repository/plagiarism/PlagiarismCaseRepository.java:
```diff
@@ -35,6 +36,15 @@ public interface PlagiarismCaseRepository extends JpaRepository<PlagiarismCase,
+            LEFT JOIN FETCH plagiarismSubmissions.plagiarismComparison plagiarismComparison
+            WHERE plagiarismCase.exercise.exerciseGroup.exam.id = :examId
+            """)
+    List<PlagiarismCase> findByExamIdWithPlagiarismSubmissionsAndComparison(@Param("examId") Long examId);
```
:bulb: Add `Post` to method name as well

akesfeden replied:
The version of this method for Course ([just above this one](https://github.com/ls1intum/Artemis/pull/5620/files/3671d7574d8d9c3a514af403d21939aaea4a6712#diff-30868ed17409051376cbb46d6aade4d5a85f41b6661b8b7a9e5d0483dd3e3f8cR30-R37)) doesn't have `Post` in its name either so I wanted to keep them similar. Should we change both of them instead?

jpbernius commented on src/main/java/de/tum/in/www1/artemis/web/rest/plagiarism/PlagiarismCaseResource.java:
```diff
@@ -66,6 +72,31 @@ public ResponseEntity<List<PlagiarismCase>> getPlagiarismCasesForCourseForInstru
+    @PreAuthorize("hasRole('INSTRUCTOR')")
+    public ResponseEntity<List<PlagiarismCase>> getPlagiarismCasesForExamForInstructor(@PathVariable long courseId, @PathVariable long examId) {
+        log.debug("REST request to get all plagiarism cases for instructor in exam with id: {}", examId);
+        Course course = courseRepository.findByIdElseThrow(courseId);
```
You can skip this database call, as this Object seems to be fetched by the other database query as well.

akesfeden replied:
That would allow a user who does not have access to the course to get knowledge about whether the plagiarism cases are empty or not. But then they can always try other `examId`s if they are instructors in at least one course so saving the database query may be more important here. Whet do you think?

jpbernius replied:
Fair enough. We should not leak if there are plagiarism cases or not. We could do something like this:
```
        var plagiarismCases = plagiarismCaseRepository.findByExamIdWithPlagiarismSubmissionsAndComparison(examId);
        Course course;
        if (!plagiarismCases.isEmpty()) {
            course = plagiarismCase.getExercise().getExerciseGroup().getExam().getCourse();
        } else {
            course = courseRepository.findByIdElseThrow(courseId);
        }
        authenticationCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.INSTRUCTOR, course, null);
```
However that is a bit more complicated. Not sure what is better.

akesfeden replied:
Then maybe we can keep the current version for simplicity since we are using `findById`, it is highly likely that the relevant course is in Hibernate cache and it won't hit the database (I was able to confirm this locally).

jpbernius commented on src/main/java/de/tum/in/www1/artemis/web/rest/plagiarism/PlagiarismCaseResource.java:
```diff
@@ -147,6 +178,31 @@ public ResponseEntity<PlagiarismCaseInfoDTO> getPlagiarismCaseForExerciseForStud
+        var user = userRepository.getUserWithGroupsAndAuthorities();
+        authenticationCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.STUDENT, course, user);
+
+        List<PlagiarismCase> plagiarismCasePerExerciseList = plagiarismCaseRepository.findByStudentIdAndExerciseIdsWithPost(user.getId(), exerciseIds);
```
I think this method does not check if the given exercise IDs belong to the course used for access check above.

akesfeden replied:
At the end of the day, we need to add some ways to always allow the all students who meet those two conditions below ([at here](https://github.com/ls1intum/Artemis/pull/5620/files/3671d7574d8d9c3a514af403d21939aaea4a6712#diff-30868ed17409051376cbb46d6aade4d5a85f41b6661b8b7a9e5d0483dd3e3f8cR61-R62)) to access these plagiarism cases to make their defence.
```
...
AND plagiarismCase.student.id = :userId
AND p.id IS NOT NULL   // p is Post
```

So I am not sure the complexity increase added by the course check gives us any benefit. (I mentioned complexity increase because we would need to check whether this is a course plagiarism case or exam plagiarism case and if so get the course from exam and so on). Instead, maybe I can just remove the course parameter from this request completely to not give a false sense of security. What do you think?

akesfeden replied:
Well, it does not look that much complicated as I initially expected, this seems to work locally, how does it look?
```java
@GetMapping("courses/{courseId}/plagiarism-cases")
@PreAuthorize("hasRole('USER')")
public ResponseEntity<Map<Long, PlagiarismCaseInfoDTO>> getPlagiarismCasesForExercisesForStudent(@PathVariable long courseId,
        @RequestParam(name = "exerciseId") Set<Long> exerciseIds) {
    log.debug("REST request to all plagiarism cases for student and exercises with ids: {}", exerciseIds);
    Course course = courseRepository.findByIdElseThrow(courseId);
    var user = userRepository.getUserWithGroupsAndAuthorities();
    authenticationCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.STUDENT, course, user);

    List<PlagiarismCase> plagiarismCasePerExerciseList = plagiarismCaseRepository.findByStudentIdAndExerciseIdsWithPost(user.getId(), exerciseIds);
    for (PlagiarismCase plagiarismCase : plagiarismCasePerExerciseList) {
        var plagiarismCaseCourse = plagiarismCase.getExercise().getCourseViaExerciseGroupOrCourseMember();
        if (!plagiarismCaseCourse.getId().equals(courseId)) {
            throw new ConflictException("Plagiarism case with id " + plagiarismCase.getId() + " is not related to the given course id " + courseId, ENTITY_NAME, "courseMismatch");
        }
    }
    Map<Long, PlagiarismCaseInfoDTO> plagiarismCaseInfoDTOs = plagiarismCasePerExerciseList.stream()
            // the following line is already checked in the SQL statement, but we want to ensure it 100%
            .filter(plagiarismCase -> plagiarismCase.getPost() != null).collect(Collectors.toMap(plagiarismCase -> plagiarismCase.getExercise().getId(),
                    plagiarismCase -> new PlagiarismCaseInfoDTO(plagiarismCase.getId(), plagiarismCase.getVerdict())));

    return ResponseEntity.ok(plagiarismCaseInfoDTOs);
}
```

akesfeden replied:
Fixed as written above in https://github.com/ls1intum/Artemis/pull/5620/commits/ce3f41d456abb3db93f8eebebf6a8bc8d970f158
